### ISSUE TITLE
Single Tap on TextField does not display Toolbar

### ISSUE THREAD

**Issue by Schoonology** [flutter/flutter#48434]
Internal: b/173627308

During final review of an app in the process of being launched, we found a surprising issue: If I, the developer, want to paste into a TextField, I long-press the field, and the toolbar appears. If the producer wants to paste into a TextField, they tap the already-focused Field...and nothing happens.

On every other app we could find, that tap presents the toolbar. As far as I can see from the documentation, Flutter's code, pub.dev, StackOverflow, and Google, there is no current issue tracking this behaviour, method providing this behaviour, or explanation of how to do so.

What I'm looking for is the ability to react to a tap on an already-focused TextField, and present the toolbar (e.g. "Paste", "Select All").

- Is there a way to do this currently that I missed? (If so, _great!_ It's entirely likely I just failed to find it, and it could even be "obvious".)
- If not, what is the best way to go about adding that? We've decided upon Flutter as our platform of choice, and I'd like to start contributing as we transition from "launch our first app" to "maintain all our apps".

**Comment by iapicca** [comment]
Hi @Schoonology 
can you please provide your `flutter doctor -v` ?
Thank you

**Comment by Schoonology** [comment]
Sure.

```
[✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Mac OS X 10.14.6 18G95, locale en-US)
    • Flutter version 1.12.13+hotfix.5 at /Users/schoon/Code/community/flutter
    • Framework revision 27321ebbad (4 weeks ago), 2019-12-10 18:15:01 -0800
    • Engine revision 2994f7e1e6
    • Dart version 2.7.0


[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/schoon/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3, Build version 11C29
    • CocoaPods version 1.8.3

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 40.2.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] VS Code (version 1.41.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.7.1

[✓] Connected device (1 available)
    • iPhone 8 • 4EE78652-B1E5-4F6E-9671-95CB1F995FC3 • ios • com.apple.CoreSimulator.SimRuntime.iOS-13-3 (simulator)

• No issues found!
```

**Comment by HansMuller** [comment]
CC @justinmc

**Comment by chunhtai** [comment]
I do not quite understand the issue. The current implementation that show toolbars upon long press and double tap. Single tap does not bring up toolbar. If you are looking for reacting to single tap, does this work for you? https://api.flutter.dev/flutter/material/TextField/onTap.html

**Comment by HansMuller** [comment]
Native iOS textfields do appear to (somewhat unreliably) show the text selection toolbar when an already-focused text field is tapped. If that's what iOS users expect we should probably implement it; we haven't done so yet though.

CC @LongCatIsLooong

**Comment by justinmc** [comment]
I tried checking which gestures show the selection menu on each platform (on physical devices).  A ✅ means that it shows the menu, and a ❌ means it does not show.  In all cases, I had something copied to the clipboard.

| Gesture | iOS | Flutter on iOS | Android | Flutter on Android |
| -------- | --- | ----------- | -------- | --------------- |
| tap         | ✅ | ❌ | ❌ | ❌ |
| double tap | ✅ | ✅ | ❌  | ✅  |
| long press | ✅ | ✅ | ✅  | ✅  |

It seems like Flutter should match each platform's behavior (regardless of whether TextField or CupertinoTextField is used).  @Schoonology would that solve your problem?

**Comment by Schoonology** [comment]
@justinmc It seems like that will completely solve my technical problem.

I'm happy to move this conversation somewhere else, but I'd like to use this as an opportunity to start contributing to Flutter. (I've done this with other platforms, but am not set up to do so here.) What does that require? Is this a good issue to start with?

**Comment by justinmc** [comment]
@Schoonology That's great!  This is not a bad issue to start with, and I'm happy to help.

Start by looking at the [contributor guide](https://github.com/flutter/flutter/blob/master/CONTRIBUTING.md), especially ["setting up your framework development evironment"](https://github.com/flutter/flutter/wiki/Setting-up-the-Framework-development-environment).

The part of the source code that you'll want to start looking at is the calls to showToolbar in text_selection.dart ([example](https://github.com/flutter/flutter/blob/23a3d1014faffa3d1a87104f1b4c90f95daa1ce7/packages/flutter/lib/src/widgets/text_selection.dart#L1030)).  Notice that TextSelection is also overridden in [material/text_selection.dart](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/text_selection.dart) and [cupertino/text_selection.dart](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/cupertino/text_selection.dart).

Feel free to reach out to me if you need any pointers.  I'm usually online on the Flutter discord channel (https://discordapp.com/invite/BS8KZyg) and I think my email is in my Github profile.

**Comment by TahaTesser** [comment]
Same happens with `CupertinoTextField`

**Comment by CiriousJoker** [comment]
@chunhtai 
You brought up reacting to onTap events as a possible solution. How do you open the toolbar manually when a specific event happens?

**Comment by chunhtai** [comment]
@CiriousJoker If you want to bring up the built-in tool bar, I don't think it is possible with current API. The problem is you don't have access to the underneath EditableText to call showToolBar because the TextFieldState is private class(Although you can manually patch it). You will have to use editable text directly, and it is very cumbersome. It is probably make more sense to wait for the fix.

**Comment by CiriousJoker** [comment]
What exactly does the fix include? Does it directly expose the showToolbar() function or does it patch the default behavior?
I specifically want to open it manually, I don't care about the default behavior at all.

I want to show the paste controls when tapping on the empty text field and select everything + copy/paste when tapping when there's existing content.

**Comment by chunhtai** [comment]
According to previous comments, I think the plan is to patch the default behavior. @justinmc might be able to confirm.

For your use case, I would probably use editable text directly.

**Comment by justinmc** [comment]
I hope to fix the show conditions in my comment https://github.com/flutter/flutter/issues/48434#issuecomment-572851932 by calling to showToolbar when the corresponding event happens.  I don't plan to directly expose showToolbar on TextField.  @CiriousJoker could you open a separate issue about being able to programmatically show the text selection menu?

**Comment by TahaTesser** [comment]
<details> 
<summary>flutter doctor -v</summary> 

``` 
[✓] Flutter (Channel dev, 1.20.0-2.0.pre, on Mac OS X 10.15.5 19F101, locale
    en-GB)
    • Flutter version 1.20.0-2.0.pre at /Users/tahatesser/Code/flutter_dev
    • Framework revision 15a28159bc (10 days ago), 2020-06-23 04:52:58 -0700
    • Engine revision 91a63d6a44
    • Dart version 2.9.0 (build 2.9.0-19.0.dev 7e72c9ae7e)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.46.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.1

 
[✓] Connected device (5 available)            
    • SM M305F      • 32003c30dc19668f          • android-arm64  • Android 10
      (API 29)
    • Taha’s iPhone • 00008020-001059882212002E • ios            • iOS 13.5.1
    • macOS desktop • macos                     • darwin-x64     • Mac OS X
      10.15.5 19F101
    • Web Server    • web-server                • web-javascript • Flutter Tools
    • Chrome        • chrome                    • web-javascript • Google Chrome
      83.0.4103.116

• No issues found!
``` 

</details>

**Comment by justinmc** [comment]
I put some effort in trying to write one-off fixes for a few of the inconsistencies that I called out in my [comment above](https://github.com/flutter/flutter/issues/48434#issuecomment-572851932) in a [branch](https://github.com/justinmc/flutter/tree/text-selection-menu-appears-when), but I think it's going to take a bigger refactor to get all of this right.  The code is getting a bit convoluted when several different files need to be able to decide when the text selection menu shows or does not show.

**Comment by Schoonology** [comment]
@justinmc That's good to hear. I took a few cracks at making this work, too, but had a lot of difficulty getting the desired behaviour without causing other issues.

**Comment by justinmc** [comment]
Yeah exactly.  There's no clear place where you can write something like "for Material text selection, when a double tap happens, do this".  The logic is spread between the 3 text_selection.dart files plus EditableText and TextField/CupertinoTextField.  I think we need to refactor it so that material/text_selection.dart and cupertino/text_selection.dart are able to have full control over deciding when their text selection menu shows up.

**Comment by dan-leech** [comment]
I've found the place but it's too complicated to make PR to Flutter right. Maybe I'm wrong.
The place is     https://github.com/flutter/flutter/blob/804c3ff31f6e517799bdd262c6b6cbe6eafac793/packages/flutter/lib/src/material/text_field.dart#L105

you need add after:
```
if(_state._effectiveFocusNode.hasPrimaryFocus && delegate.selectionEnabled && shouldShowSelectionToolbar) {                
    editableText.showToolbar();
}
```

I'll appreciate if someone with all fits push this fix.

**Comment by justinmc** [comment]
If I make that exact change, the toolbar doesn't hide with subsequent taps.  Even if I change it to `.toggleToolbar()`.  Maybe there's some smallish change that will work, but really this needs a deeper refactor.

**Comment by dan-leech** [comment]
@justinmc My working solution looks like now:
```
   if (delegate.selectionEnabled) {
      switch (Theme.of(_state.context).platform) {
        case TargetPlatform.iOS:
        case TargetPlatform.macOS:
          switch (details.kind) {
            case PointerDeviceKind.mouse:
            case PointerDeviceKind.stylus:
            case PointerDeviceKind.invertedStylus:
              // Precise devices should place the cursor at a precise position.
              renderEditable.selectPosition(cause: SelectionChangedCause.tap);
              break;
            case PointerDeviceKind.touch:
            case PointerDeviceKind.unknown:
              // On macOS/iOS/iPadOS a touch tap places the cursor at the edge
              // of the word.
              renderEditable.selectWordEdge(cause: SelectionChangedCause.tap);

              if (_state._effectiveFocusNode.hasPrimaryFocus &&
                  delegate.selectionEnabled &&
                  shouldShowSelectionToolbar) {
                editableText.showToolbar();
              }
              break;
          }
          break;
```

**Comment by dan-leech** [comment]
Why this is closed?

The problem exists.
All duplicates are closed too.

**Comment by justinmc** [comment]
This issue is currently open as it should be.

I'm currently working on a refactor that should alleviate some of the problems with the codebase that I mentioned [above](https://github.com/flutter/flutter/issues/48434#issuecomment-656757943).  After that I think we'll be able to address the inconsistencies from my [table above](https://github.com/flutter/flutter/issues/48434#issuecomment-572851932).

**Comment by supposedlysam-bb** [comment]
@justinmc any notable update on this? I'd like to hear any progress you've been able to make. I believe iOS users are still experiencing some friction interacting with text fields because of this bug. Thank you for your efforts so far!

**Comment by justinmc** [comment]
This is the PR that will unblock the clean fix of this and other bugs: https://github.com/flutter/flutter/pull/84232 CC @LongCatIsLooong 

Once we are able to handle tap interactions via Actions, then I think we'll be able to untangle problems like this by creating one place where the the logic for all platforms lives.

**Comment by bdlukaa** [comment]
Any updates on this?

**Comment by justinmc** [comment]
We are in the middle of refactoring so that we can handle things like this using the Actions system.  It should make this much cleaner and less prone to errors.

It could be awhile until that is done, though.  If anyone wants to write an interim fix in the meantime feel free.

**Comment by doppio** [comment]
Is there any sort of workaround to get a single tap to bring up the toolbar on iOS? Using `CupertinoTextField` doesn't seem to work.

My app's primary UI flow involves pasting text into a field. It's in beta, and multiple testers have already reported that they can't paste by tapping on the field. I'm worried about releasing to production when the primary interaction feels wrong on iOS.

Any thoughts on how I can work around this until the real fix is possible?
(cc @justinmc)

**Comment by justinmc** [comment]
No workaround that I'm aware of, unfortunately.  I'm open to reviewing a PR for this if anyone vitally needs this without waiting for the refactor.  We're making progress on the refactor but it's a big change.

**Comment by doppio** [comment]
I created a PR for this over at #95644 if you're willing to take a look at some point @justinmc. :) It's my first contribution and I'm not sure if I did it right though! 🤞

**Comment by payam-zahedi** [comment]
@justinmc 
Can we have `onToolbarOptionChange` callback? This will help us To detect the changes of the toolbar option.
Imagine that we have custom autocomplete, and we want to disappear our overlay when the toolbar has been opened.
This is what exactly we need in our app.

**Comment by andaas4** [comment]
On my Android (Samsung S21) device the toolbar is shown after a single tap on native apps when the text field is already focused.

**Comment by 0n9aldrin** [comment]
> Is there any sort of workaround to get a single tap to bring up the toolbar on iOS? Using `CupertinoTextField` doesn't seem to work.
> 
> My app's primary UI flow involves pasting text into a field. It's in beta, and multiple testers have already reported that they can't paste by tapping on the field. I'm worried about releasing to production when the primary interaction feels wrong on iOS.
> 
> Any thoughts on how I can work around this until the real fix is possible? (cc @justinmc)

Did you end up finding a fix for this?

**Comment by Renzo-Olivares** [comment]
This was fixed by #111995.

**Comment by DmytroMeleshko** [comment]
Please help me, maybe I'm doing something wrong.
I switched to the master channel to test opening a toolbar with a single tap, but it didn't work for me both on Android and iOS. The toolbar still only opens with a double tap or long press.
